### PR TITLE
Add the ecms api publiser module and custom entity

### PR DESCRIPTION
**Pull request by pfrilling**
## Summary
This adds the `ecms_api_publisher` custom module and currently installs a new entity type that will keep track of the sites to publish content to. This is mainly the custom entity and the supporting files to allow these entities to be created in the admin area. Functional and Unit tests have also been provided.

## Metadata
| Question | Answer |
|----------|--------|
| Did you use a meaningful pull request title? | yes
| Did you apply meaningful labels to the pull request? | yes
| Did you perform a self review first? | yes
| Documentation reflects changes? | yes
| `CHANGELOG` reflects changes? | yes
| Unit/Functional tests reflect changes? | yes
| Did you perform browser testing? | yes
| Risk level | Medium
| Relevant links | [RIG-57](https://thinkoomph.jira.com/browse/rig-57)

### CODE REVIEW

File: ecms_base/modules/custom/ecms_api_publisher/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# eCMS API Publisher
+
+This module manages publishing content to other sites within the eCMS network.
+It does this using the JSON API and a custom entity type.
+
+## ecms_api_site Custom entity
+The eCMS APi Site custom entity stores the endpoints to publish content to and
```
Comment (ben-hamelin) [inline ecms_base/modules/custom/ecms_api_publisher/README.md:7]: nitpick "API"

File: ecms_base/modules/custom/ecms_api_publisher/src/EcmsApiSiteAccessControlHandler.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+   * {@inheritdoc}
+   */
+  protected function checkAccess(EntityInterface $entity, $operation, AccountInterface $account): AccessResult {
+    switch ($operation) {
+      case 'view':
+        // Check if the entity belongs to the user.
+        if ($account->id() === $entity->getOwnerId()) {
+          return AccessResult::allowedIfHasPermission($account, 'view own published ecms api site entities');
+        }
+
```
Comment (ben-hamelin) [inline ecms_base/modules/custom/ecms_api_publisher/src/EcmsApiSiteAccessControlHandler.php:29]: Should you add 'break' statements to these? Default case?

File: ecms_base/modules/custom/ecms_api_publisher/src/Form/EcmsApiSiteForm.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * Form controller for eCMS API Site entities.
+ *
+ * @ingroup ecms_api_publisher
+ */
+class EcmsApiSiteForm extends ContentEntityForm {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
```
Comment (ben-hamelin) [inline ecms_base/modules/custom/ecms_api_publisher/src/Form/EcmsApiSiteForm.php:20]: return type

File: ecms_base/modules/custom/ecms_api_publisher/src/Form/EcmsApiSiteForm.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    $form = parent::buildForm($form, $form_state);
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function save(array $form, FormStateInterface $form_state) {
```
Comment (ben-hamelin) [inline ecms_base/modules/custom/ecms_api_publisher/src/Form/EcmsApiSiteForm.php:29]: return type

File: ecms_base/modules/custom/ecms_api_publisher/tests/src/Functional/InstallationTest.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  /**
+   * Define the additional modules to install.
+   *
+   * @var string[]
+   */
+  protected static $modules = ['ecms_api_publisher'];
+
+  /**
+   * Test the ecms_api_recipient installation.
```
Comment (ben-hamelin) [inline ecms_base/modules/custom/ecms_api_publisher/tests/src/Functional/InstallationTest.php:39]: should be "ecms_api_publisher"

File: ecms_base/modules/custom/ecms_api_publisher/tests/src/Unit/Entity/EcmsApiSiteTest.php [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+    // Assert the custom fields are returned.
+    foreach (self::ENTITY_BASE_FIELDS as $key => $value) {
+      $this->assertEquals($fields[$key], $results[$key]);
+    }
+  }
+
+  /**
+   * Setup the plugin manager mocked methods.
+   */
+  protected function setupPluginManager(): void {
```
Comment (ben-hamelin) [inline ecms_base/modules/custom/ecms_api_publisher/tests/src/Unit/Entity/EcmsApiSiteTest.php:425]: seeing return types that are non-void in this method

File: ecms_base/modules/custom/ecms_api_publisher/tests/src/Unit/Entity/EcmsApiSiteTest.php [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+      ->willReturn($value);
+  }
+
+  /**
+   * Build class with a number fake field methods.
+   *
+   * @param mixed $value
+   *   Value to return.
+   */
+  protected function buildFakeField($value) {
```
Comment (ben-hamelin) [inline ecms_base/modules/custom/ecms_api_publisher/tests/src/Unit/Entity/EcmsApiSiteTest.php:474]: return type

**Review by ben-hamelin — COMMENTED** [review]
Ok noted a few thinsgs

**Comment by pfrilling** [comment]
@ben-hamelin, I made the changes you suggested and rebased everything. Can you give this another look?

**Comment by ben-hamelin** [comment]
Ok LGTM!
/approved

**Pull request merged**